lxd/storage/drivers/driver_powerflex_utils.go

```diff
@@ -0,0 +1,1021 @@
+func (d *powerflex) getHostNQN() (string, error) {
+	uuidBytes, err := os.ReadFile(shared.VarPath("uuid"))
+	if err != nil {
+		return "", fmt.Errorf("Failed to read UUID file: %w", err)
+	}
+
+	return fmt.Sprintf("nqn.2014-08.org.nvmexpress:uuid:%s", string(uuidBytes)), nil
+}
+
+// getHostname returns the hostname of this host.
+func (d *powerflex) getHostname() (string, error) {
+	hostname, err := os.Hostname()
+	if err != nil {
+		return "", fmt.Errorf("Failed to get hostname: %w", err)
+	}
+
+	return hostname, nil
+}
+
+// getVolumeType returns the selected provisioning type of the volume.
+func (d *powerflex) getVolumeType(vol Volume) string {
+	if vol.config["block.type"] != "" {
+		return vol.config["block.type"]
+	}
+
+	return "thin"
+}
+
+// getVolumeName returns the fully qualified name derived from the volume.
+func (d *powerflex) getVolumeName(vol Volume, snapName string) string {
+	var out string
+	parentName, snapshotName, isSnapshot := api.GetParentAndSnapshotName(vol.name)
+
+	// Only use filesystem suffix on filesystem type image volumes (for all content types).
+	// Since PowerFlex volumes can be 31 characters long at max, use the image fingerprint.
+	if vol.volType == VolumeTypeImage {
+		parentName = fmt.Sprintf("%s_%s", parentName[0:12], vol.ConfigBlockFilesystem())
+	}
+
+	if vol.contentType == ContentTypeBlock {
+		parentName = fmt.Sprintf("%s%s", parentName, powerFlexBlockVolSuffix)
+	} else if vol.contentType == ContentTypeISO {
+		parentName = fmt.Sprintf("%s%s", parentName, powerFlexISOVolSuffix)
+	}
+
+	// Use volume's type as storage volume prefix, unless there is an override in powerFlexVolTypePrefixes.
+	volumeTypePrefix := string(vol.volType)
+	volumeTypePrefixOverride, foundOveride := powerFlexVolTypePrefixes[vol.volType]
+	if foundOveride {
+		volumeTypePrefix = volumeTypePrefixOverride
+	}
+
+	if snapName != "" {
+		// Always use the provided snapshot name if specified.
+		out = fmt.Sprintf("%s_%s@%s", volumeTypePrefix, parentName, snapName)
+	} else {
+		if isSnapshot {
+			// If volumeName is a snapshot (<vol>/<snap>) and snapName is not set,
+			// assume that it's a normal snapshot (not a zombie) and prefix it with
+			// "snapshot_".
+			out = fmt.Sprintf("%s_%s@snapshot_%s", volumeTypePrefix, parentName, snapshotName)
+		} else {
+			out = fmt.Sprintf("%s_%s", volumeTypePrefix, parentName)
+		}
+	}
+
+	return out
+}
+
+// createNVMeHost creates this NVMe host in PowerFlex.
+// The operation is idempotent and locked using lock name powerflex.host.
+func (d *powerflex) createNVMeHost() (string, error) {
+	nqn, err := d.getHostNQN()
+	if err != nil {
+		return "", err
+	}
+
+	unlock := locking.Lock(d.state.ShutdownCtx, "powerflex.host")
+	defer unlock()
+
+	var hostID string
+
+	host, err := d.powerFlex().getNVMeHostByNQN(nqn)
+	if err != nil {
+		if !api.StatusErrorCheck(err, http.StatusNotFound) {
+			return "", err
+		}
+
+		hostname, err := d.getHostname()
```

Comment:
I'd like to understand more about the role server hostname plays here.
Normally we would use d.state.ServerName or falling back to host name if "none"

Reply:
When creating a host entry in PowerFlex you can optionally provide a name for it. Otherwise PowerFlex will pick one derived from the ID that gets generated for this new host. 

If I remember correctly I moved away from `state.ServerName` since it didn't bring the required value. From the comments I see the value is only set in case LXD is clustered. I will modify `getHostname()` to prefer the value from `state`.

Reply:
lets call it getServerName() and it call fall back to hostname internally. whereas getHostname suggests it always returns hostname

Reply:
The function `getServerName()` now tries to read from the state and falls back to the actual hostname.